libsimulator/src/SocialForceModel.cpp

```diff
@@ -0,0 +1,182 @@
+
+        if(model.radius >= distance) {
+            throw SimulationError(
+                "Model constraint violation: Agent {} too close to agent {}: distance {}, "
+                "radius {}",
+                agent.pos,
+                neighbor.pos,
+                distance,
+                model.radius);
+        }
+    }
+    const auto maxRadius = model.radius / 2;
+    const auto lineSegments = geometry.LineSegmentsInDistanceTo(maxRadius, agent.pos);
+    if(std::begin(lineSegments) != std::end(lineSegments)) {
+        throw SimulationError(
+            "Model constraint violation: Agent {} too close to geometry boundaries, distance <= "
+            "{}/2",
+            agent.pos,
+            model.radius);
+    }
+}
+
+Point SocialForceModel::drivingForce(const GenericAgent& agent) const
+{
+    auto& model = std::get<SocialForceModelData>(agent.model);
+    Point e0 = (agent.destination - agent.pos).Normalized();
+    return (e0 * model.desiredSpeed - model.velocity) / model.reactionTime;
+};
+double SocialForceModel::pushing_Force(double A, double B, double r, double distance) const
+{
+    return A * exp((r - distance) / B);
+}
+
+Point SocialForceModel::AgentForce(const GenericAgent& ped1, const GenericAgent& ped2) const
+{
+    double dist = (ped1.pos - ped2.pos).Norm();
+    auto& model1 = std::get<SocialForceModelData>(ped1.model);
+    auto& model2 = std::get<SocialForceModelData>(ped2.model);
+
+    double total_radius = model1.radius + model2.radius;
+    double pushing_force_length =
+        pushing_Force(model1.agentScale, model1.forceDistance, total_radius, dist);
+    double friction_force_length = 0;
+    Point n_ij = (ped1.pos - ped2.pos).Normalized();
+    Point tangent = n_ij.Rotate90Deg();
+
+    if(dist < total_radius) {
+        pushing_force_length += this->bodyForce * (total_radius - dist);
+        friction_force_length += this->friction * (total_radius - dist) *
+                                 ((model2.velocity - model1.velocity).ScalarProduct(tangent));
+    }
+
+    return n_ij * pushing_force_length + tangent * friction_force_length;
+};
+
+Point SocialForceModel::ObstacleForce(const GenericAgent& agent, const LineSegment& segment) const
+{
+    auto& model = std::get<SocialForceModelData>(agent.model);
+    Point pt = segment.ShortestPoint(agent.pos);
+    double dist = (agent.pos - pt).Norm();
+    double pushing_force_length =
+        pushing_Force(model.obstacleScale, model.forceDistance, model.radius, dist);
+    double friction_force_length = 0;
+    Point n_ij = (agent.pos - pt).Normalized();
+    Point tangent = n_ij.Rotate90Deg();
+
+    if(dist < model.radius) {
+        pushing_force_length += this->bodyForce * (model.radius - dist);
+        friction_force_length +=
+            this->friction * (model.radius - dist) * (model.velocity.ScalarProduct(tangent));
+    }
+    return n_ij * pushing_force_length + tangent * friction_force_length;
```

Comment:
This seems to be duplicated (see `AgentForce`). Is there a reason for not moving this to a separate function?

Reply:
the force calculation was extracted into new function `ForceBetweenPoints`